### ISSUE TITLE
Update discord.py library

### ISSUE THREAD

**Issue by EndenDragon** [EndenDragon/Anankos#1]
Thus removing dependency for discord-py-slash-commands and support threads better.

**Comment by EndenDragon** [comment]
# Discord.py 1.7.2 → 2.7.1 Upgrade Guide

## Overview

This guide documents every breaking change between the current codebase (discord.py 1.7.2 + discord-py-slash-command) and discord.py 2.7.1, with concrete before/after code derived from the actual source files. Changes are ordered by priority.

### The Short Version

Discord.py 2.7.1 natively supports everything that previously required `discord-py-slash-command` and the monkey patch:
- Buttons and components → `discord.ui.View`, `discord.ui.Button`
- Slash commands → `discord.app_commands`
- Polls → `message.poll` (native since 2.4)

All three external workarounds (`discord-py-slash-command`, `monkey_patch.py`, raw `_state.http` calls) can be replaced with cleaner first-party APIs.

---

## Part 1: Dependencies (`requirements.txt`)

### Changes required

```diff
- discord.py==1.7.2
+ discord.py==2.7.1
- git+https://github.com/discord-py-slash-commands/discord-py-slash-command.git@6a5cb290bd4ae3b3d172072094869f0fa4652f32
- pycrypto==2.6.1
+ pycryptodome==3.20.0
  aiosqlite==0.15.0
  urlextract==1.0.0
  python-twitter==3.5
  Unidecode==1.1.1
  asyncpraw==7.5.0
  python-dateutil==2.8.2
  markdownify==0.11.6
  beautifulsoup4==4.12.2
  ffmpeg-python==0.2.0
```

**Notes:**
- `discord-py-slash-command` is fully replaced by native discord.py 2.x APIs — remove it entirely.
- `pycrypto` is an abandoned package with known security vulnerabilities. Replace with `pycryptodome`, which is a drop-in replacement. The `from Crypto.Cipher import AES` import in `priconne_notification.py` does **not** change — `pycryptodome` keeps the `Crypto` namespace for compatibility.
- `python-twitter` (used by `image_embed.py`) is unrelated to the discord.py upgrade but is worth noting: it targets the v1 Twitter/X API which has been heavily restricted. This may already be non-functional.

---

## Part 2: `monkey_patch.py` — Delete entirely

**File: `Anankos/monkey_patch.py`**

```python
# DELETE THIS FILE
```

The monkey patch existed to extend `discord-py-slash-command`'s `ComponentMessage` with a `poll` attribute, because neither discord.py 1.x nor the slash command library had native poll support. Discord.py 2.4+ adds `message.poll` natively. With the slash command library gone, `ComponentMessage` no longer exists, so this file has no purpose.

Also remove the two lines that reference it in `bot.py`:

```python
# Remove these two lines from bot.py:
from Anankos.monkey_patch import monkey_patch
monkey_patch()
```

---

## Part 3: `run.py` — Update entry point

**File: `run.py`**

Discord.py 2.x deprecates synchronous `client.run()` as the only entry point. The recommended pattern uses an async context manager, which ensures proper cleanup of aiohttp sessions and other async resources.

```python
# Before (run.py)
from Anankos.bot import Anankos
from config import CONFIG

anankos = Anankos(CONFIG)
anankos.run(CONFIG["bot_token"])
```

```python
# After (run.py)
import asyncio
from Anankos.bot import Anankos
from config import CONFIG

async def main():
    async with Anankos(CONFIG) as anankos:
        await anankos.start(CONFIG["bot_token"])

asyncio.run(main())
```

> **Note:** `client.run()` still works in 2.7.1 as a convenience method, so this is not a hard blocker. But the async context manager pattern is the correct approach and avoids deprecation warnings.

---

## Part 4: `bot.py` — Core bot restructure

**File: `Anankos/bot.py`**

This file requires the most structural changes.

### 4.1 Remove `discord-py-slash-command` and monkey patch imports

```python
# Remove these lines entirely:
from discord_slash import SlashCommand
from Anankos.monkey_patch import monkey_patch

monkey_patch()
```

### 4.2 Switch from `discord.Client` to `commands.Bot`

The bot currently subclasses `discord.Client` directly. Discord.py 2.x's app command system (`bot.tree`) works with both `Client` and `commands.Bot`, but `commands.Bot` wires up `CommandTree` automatically and is the lower-friction path. Since this bot doesn't currently use prefix commands via the commands framework, the change is minimal.

```python
# Before
import discord

class Anankos(discord.Client):
    def __init__(self, config):
        super().__init__(
            activity=discord.Game(name=config.get("playing_status", "with Dragon Veins")),
            intents=discord.Intents.all()
        )
        self.slash_command = SlashCommand(self)
        # ... module instantiation
```

```python
# After
import discord
from discord.ext import commands

class Anankos(commands.Bot):
    def __init__(self, config):
        super().__init__(
            command_prefix=config.get("cmd_prefix", "!"),
            activity=discord.Game(name=config.get("playing_status", "with Dragon Veins")),
            intents=discord.Intents.all()
        )
        self.config = config
        self.db = None
        self.cmd_prefix = config.get("cmd_prefix", "!")
        # Do NOT start background tasks here — move to setup_hook()
        self.bad_words = BadWords(self, config.get("bad_words", []))
        self.pick_a_number = PickANumber(self, ...)
        # ... rest of module instantiation (without background task modules)
```

### 4.3 Add `setup_hook()` — replaces `on_connect()` for initialization

`setup_hook()` is called after login but before the gateway connection opens. It is the correct place for async initialization, database setup, background task starts, and persistent view registration. `on_connect()` can be removed.

```python
# Before: initialization split across __init__ and on_connect
async def on_connect(self):
    if self.db is None:
        self.db = await aiosqlite.connect("db.sqlite3", detect_types=sqlite3.PARSE_DECLTYPES|sqlite3.PARSE_COLNAMES)
        await self.pick_a_number.create_tables()
        await self.trivia.create_tables()
        await self.art_mention.create_tables()
```

```python
# After: consolidate into setup_hook()
async def setup_hook(self):
    # Database
    self.db = await aiosqlite.connect("db.sqlite3", detect_types=sqlite3.PARSE_DECLTYPES|sqlite3.PARSE_COLNAMES)
    await self.pick_a_number.create_tables()
    await self.trivia.create_tables()
    await self.art_mention.create_tables()

    # Register persistent views so buttons work after restart
    self.add_view(ArtMentionView(self.art_mention))

    # Start background tasks
    self.trivia.start_tasks()
    self.art_mention.start_tasks()
    self.dragalia_notification.start_tasks()
    self.reddit_publish.start_tasks()
    self.priconne_notification.start_tasks()
```

### 4.4 Replace `self.loop.create_task()` in `on_message`

```python
# Before
async def on_message(self, message):
    self.loop.create_task(self.pick_a_number.on_message(message))
    self.loop.create_task(self.bad_words.on_message(message))
    # ...
```

```python
# After
import asyncio

async def on_message(self, message):
    asyncio.create_task(self.pick_a_number.on_message(message))
    asyncio.create_task(self.bad_words.on_message(message))
    # ...
```

### 4.5 Remove `on_component()` — replaced by View callbacks

```python
# Remove this handler entirely. Button interactions are now handled
# by the discord.ui.View subclass in art_mention.py.
async def on_component(self, component):
    await self.art_mention.on_component(component)
```

### 4.6 Update `on_ready` string (cosmetic)

```python
# Before
print("[Anankos by EndenDragon#1337]")

# After — discriminators no longer apply to most users
print("[Anankos by EndenDragon]")
```

---

## Part 5: Background tasks — all affected modules

**Files: `trivia.py`, `art_mention.py`, `dragalia_notification.py`, `priconne_notification.py`, `reddit_publish.py`**

All five modules create background tasks in `__init__` via `self.client.loop.create_task()`. This breaks in 2.x because:
1. The `loop` parameter is removed from the `Client`/`Bot` constructor entirely.
2. Creating tasks in `__init__` is inherently unsafe — the event loop may not be running yet.

The solution is the `discord.ext.tasks` loop decorator. Each module should expose a `start_tasks()` method called from `bot.setup_hook()`.

### Pattern to apply across all five modules

```python
# Before (example from trivia.py)
from Anankos.trivia import Trivia

class Trivia:
    def __init__(self, client, ...):
        # ...
        self.bg_task = self.client.loop.create_task(self.background_task())

    async def background_task(self):
        await self.client.wait_until_ready()
        while not self.client.is_closed():
            # ... do work
            await asyncio.sleep(120)
```

```python
# After
from discord.ext import tasks

class Trivia:
    def __init__(self, client, ...):
        # ... no task creation here

    def start_tasks(self):
        self.background_task.start()

    @tasks.loop(seconds=120)
    async def background_task(self):
        # ... do work (no while loop needed, decorator handles repetition)

    @background_task.before_loop
    async def before_background_task(self):
        await self.client.wait_until_ready()
```

### Module-specific notes

**`dragalia_notification.py`** — the background task has a custom initial delay to align to a minute boundary before entering its loop:

```python
# Before
async def background_task(self):
    await self.client.wait_until_ready()
    utc = datetime.datetime.utcnow()
    wait = 5 - utc.second if utc.second <= 5 else 60 - utc.second + 5
    await asyncio.sleep(wait)
    while not self.client.is_closed():
        utc = datetime.datetime.utcnow()
        if utc.hour == 6 and utc.minute == 0:
            await self.run_reminders()
            await asyncio.sleep(120)
        await asyncio.sleep(30)
```

```python
# After — tasks.loop handles the interval; initial alignment via before_loop
@tasks.loop(seconds=30)
async def background_task(self):
    utc = datetime.datetime.utcnow()
    if utc.hour == 6 and utc.minute == 0:
        await self.run_reminders()

@background_task.before_loop
async def before_background_task(self):
    await self.client.wait_until_ready()
    utc = datetime.datetime.utcnow()
    wait = (5 - utc.second) if utc.second <= 5 else (60 - utc.second + 5)
    await asyncio.sleep(wait)
```

**`art_mention.py`** has two tasks — both need the same treatment:

```python
# Before
self.bg_task_delete_roles = self.client.loop.create_task(self.background_task_delete_roles())
self.bg_task_update_pingboard = self.client.loop.create_task(self.background_task_update_pingboard())
```

```python
# After
def start_tasks(self):
    self.background_task_delete_roles.start()
    self.background_task_update_pingboard.start()

@tasks.loop(hours=12)
async def background_task_delete_roles(self):
    for guild in self.client.guilds:
        for role in list(guild.roles):
            if role.name.endswith(" - Fanart Notification"):
                character = role.name.split()[0]
                if await self.role_expired(character):
                    try:
                        await role.delete()
                    except:
                        pass

@tasks.loop(minutes=10)
async def background_task_update_pingboard(self):
    if not self.pingboard_uptodate:
        self.pingboard_uptodate = True
        try:
            await self.update_pingboard()
        except:
            pass

@background_task_delete_roles.before_loop
@background_task_update_pingboard.before_loop
async def before_tasks(self):
    await self.client.wait_until_ready()
```

---

## Part 6: `art_mention.py` — Button system rewrite

**File: `Anankos/art_mention.py`**

This is the largest single change. The entire component system (button creation, pingboard, interaction handling) is built on `discord-py-slash-command` and must be rewritten using `discord.ui`.

### 6.1 Remove the old import

```python
# Remove:
from discord_slash.utils.manage_components import create_actionrow, create_button, ButtonStyle
```

### 6.2 Button style mapping

| Old (`discord_slash`) | New (`discord.py 2.x`) |
|---|---|
| `ButtonStyle.blue` | `discord.ButtonStyle.primary` |
| `ButtonStyle.gray` | `discord.ButtonStyle.secondary` |
| `ButtonStyle.green` | `discord.ButtonStyle.success` |
| `ButtonStyle.red` | `discord.ButtonStyle.danger` |

### 6.3 Create a persistent `View` class

Persistent views require:
- `timeout=None` on the View
- An explicit `custom_id` on every button component
- The view registered at startup via `bot.add_view(ArtMentionView(art_mention))`

Since the buttons are dynamic (one per character), the best approach is a single persistent View class with a single `DynamicButton` that accepts any `custom_id` matching the `art_mention ` prefix.

```python
# New: add this class (can live in art_mention.py or a separate ui.py)
import discord

class ArtMentionButton(discord.ui.DynamicItem[discord.ui.Button], template=r"art_mention (?P<character>\w+)"):
    """A persistent button for subscribing/unsubscribing to a character's fanart notifications."""

    def __init__(self, character: str, style: discord.ButtonStyle = discord.ButtonStyle.primary, label: str = None):
        super().__init__(
            discord.ui.Button(
                label=label or character,
                custom_id=f"art_mention {character}",
                style=style,
                emoji="🔔",
            )
        )
        self.character = character

    @classmethod
    async def from_custom_id(cls, interaction: discord.Interaction, item: discord.ui.Button, match: re.Match):
        return cls(character=match.group("character"), label=item.label, style=item.style)

    async def callback(self, interaction: discord.Interaction):
        art_mention = interaction.client.art_mention
        character = self.character
        author = interaction.user
        existing = await art_mention.get_all_user_subscriptions(author.id)
        await interaction.response.defer(ephemeral=True)
        if character in existing:
            await art_mention.unsubscribe_user(author.id, character)
            message = f"Unsubscribed from **{character}**."
        else:
            await art_mention.subscribe_user(author.id, character)
            count = await art_mention.get_sub_count(character)
            message = f"✅ Successfully subscribed to **{character} ({count})**."
        await interaction.followup.send(message, ephemeral=True)
```

> **Note on `DynamicItem`:** `discord.ui.DynamicItem` (added in 2.4) is the right tool here because the character name is embedded in the `custom_id` and varies per button. Register it once with `bot.add_dynamic_items(ArtMentionButton)` in `setup_hook()` instead of `bot.add_view()`.

### 6.4 Replace `create_button` / `create_actionrow` in `on_message`

```python
# Before
button = create_button(style=ButtonStyle.blue, label=name, custom_id="art_mention {}".format(name), emoji="🔔")
button_list.append(button)
# ...
components = []
button_list = list(self.divide_chunks(button_list, 5))
for chunk in button_list:
    components.append(create_actionrow(*chunk))
await channel.send(mentions, mention_author=False, components=components)
```

```python
# After
view = discord.ui.View(timeout=None)
for name in names[:25]:  # Discord hard limit: 25 components per message
    style = discord.ButtonStyle.primary
    view.add_item(ArtMentionButton(character=name, style=style, label=name))
await channel.send(mentions, mention_author=False, view=view)
```

### 6.5 Replace `create_button` / `create_actionrow` in `update_pingboard`

```python
# Before
button_style = ButtonStyle.gray
# ... ratio checks ...
button = create_button(style=button_style, label="{} ({})".format(character, len(users)), custom_id="art_mention {}".format(character))
button_list.append(button)
# ... chunking into actionrows ...
await existing_messages[chunk_id].edit(components=components)
await channel.send("​", components=components)
```

```python
# After
# Build a list of (character, style, label) tuples first, then chunk into Views.
# Discord allows max 5 rows × 5 buttons = 25 buttons per message.
# Split characters into groups of 25 for multi-message pingboard.
entries = []
for character, users in subs.items():
    ratio = len(users) / max_users if max_users else 0
    if ratio > 0.55:
        style = discord.ButtonStyle.danger
    elif ratio > 0.25:
        style = discord.ButtonStyle.success
    elif ratio > 0.1:
        style = discord.ButtonStyle.primary
    else:
        style = discord.ButtonStyle.secondary
    entries.append((character, style, f"{character} ({len(users)})"))

# Chunk into messages of 25 buttons each
message_chunks = list(self.divide_chunks(entries, 25))
chunk_id = 0
for message_chunk in message_chunks:
    view = discord.ui.View(timeout=None)
    for character, style, label in message_chunk:
        view.add_item(ArtMentionButton(character=character, style=style, label=label))
    if chunk_id < len(existing_messages):
        await existing_messages[chunk_id].edit(content="​", view=view)
    else:
        await channel.send("​", view=view)
    chunk_id += 1
for message in existing_messages[chunk_id:]:
    await message.delete()
```

### 6.6 Remove `on_component` handler

```python
# Remove entirely — interaction is now handled by ArtMentionButton.callback()
async def on_component(self, component):
    custom_id = component.custom_id.split()
    if custom_id[0] != "art_mention":
        return
    # ...
```

### 6.7 Register in `bot.setup_hook()`

```python
# In bot.py setup_hook():
self.add_dynamic_items(ArtMentionButton)
```

---

## Part 7: `utils.py` — Thread creation

**File: `Anankos/utils.py`**

```python
# Before — raw HTTP via private _state API
async def create_thread(message, name, auto_archive_duration=1440):
    channel = message.channel
    data = await message.guild._state.http.request(
        discord.http.Route(
            "POST",
            "/channels/{channel_id}/messages/{message_id}/threads",
            channel_id=channel.id,
            message_id=message.id
        ),
        json={"name": name, "auto_archive_duration": auto_archive_duration}
    )
    data["position"] = 100
    channel = discord.TextChannel(state=message.guild._state, guild=message.guild, data=data)
    return channel
```

```python
# After — native API, returns a discord.Thread object
async def create_thread(message, name, auto_archive_duration=1440):
    thread = await message.create_thread(
        name=name,
        auto_archive_duration=auto_archive_duration
    )
    return thread
```

> **Note:** The return type changes from `discord.TextChannel` to `discord.Thread`. Callers use `thread.send(...)` the same way, so `art_mention.py`'s call to `await channel.send(...)` and `await self.client.image_embed.post_image_embeds(message, channel)` will work without changes — both `TextChannel` and `Thread` implement the `Messageable` interface. Verify that any code inspecting the type or accessing `TextChannel`-specific attributes still works.
>
> Valid values for `auto_archive_duration`: `60`, `1440`, `4320`, `10080` (minutes). The existing default of `1440` (1 day) is valid.

---

## Part 8: `activities.py` — Embedded activity invites

**File: `Anankos/activities.py`**

```python
# Before — raw HTTP via private _state API
async def create_embedded_invite(self, channel, target_application_id):
    data = await channel._state.http.request(
        discord.http.Route(
            "POST",
            "/channels/{channel_id}/invites",
            channel_id=channel.id
        ),
        json={
            "max_age": 0,
            "target_type": 2,
            "target_application_id": target_application_id
        }
    )
    return discord.Invite.from_incomplete(data=data, state=channel._state)
```

```python
# After — native API
async def create_embedded_invite(self, channel, target_application_id):
    invite = await channel.create_invite(
        max_age=0,
        target_type=discord.InviteTarget.embedded_application,
        target_application_id=target_application_id
    )
    return invite
```

The rest of `activities.py` (the `on_message` handler, the activity options dict) is unchanged.

---

## Part 9: `reddit_publish.py` — Crossposting and background task

**File: `Anankos/reddit_publish.py`**

### 9.1 Fix background task creation in `__init__`

```python
# Before
self.bg_task = self.client.loop.create_task(self.background_task())
```

Move to a `start_tasks()` method and use `@tasks.loop` (see Part 5 pattern). The Reddit stream loop is unusual — it's a long-running `async for` that itself never returns, so the outer `while not is_closed()` loop only retries on exception. This maps well to `@tasks.loop(seconds=120)` with error handling:

```python
from discord.ext import tasks

def start_tasks(self):
    self.background_task.start()

@tasks.loop(seconds=120)
async def background_task(self):
    try:
        subreddit = await self.reddit.subreddit("CorrinConclave")
        async for submission in subreddit.stream.submissions(skip_existing=True):
            if hasattr(submission, "created_utc") and submission.created_utc:
                if abs(time.time() - submission.created_utc) > 600:
                    print("skipped too old" + " https://reddit.com" + submission.permalink)
                    continue
            else:
                continue
            try:
                permalink = "https://reddit.com" + submission.permalink
                embed = await self.get_rich_embed(permalink)
                if embed:
                    channel = self.client.get_channel(self.source_chan_id)
                    await channel.send(embed=embed)
            except Exception as e:
                print("Reddit Publish submission error:", e)
    except Exception as e:
        print("Reddit Publish error:", e)

@background_task.before_loop
async def before_background_task(self):
    await self.client.wait_until_ready()
```

### 9.2 Fix crosspost (`publish_message`)

```python
# Before — raw HTTP via private _state API
async def publish_message(self, message):
    channel = message.channel
    await message.guild._state.http.request(
        discord.http.Route(
            "POST",
            "/channels/{channel_id}/messages/{message_id}/crosspost",
            channel_id=channel.id,
            message_id=message.id,
        )
    )
```

```python
# After — single native call
async def publish_message(self, message):
    await message.publish()
```

---

## Part 10: `role_reaction.py` — Emoji type checks

**File: `Anankos/role_reaction.py`**

`emoji.is_custom_emoji()` was not confirmed to exist in discord.py 2.7.1. `emoji.is_unicode_emoji()` was confirmed. To be safe, replace both with explicit attribute checks:

```python
# Before
if emoji.is_custom_emoji():
    lookup_str = emoji.id
if emoji.is_unicode_emoji():
    lookup_str = emoji.name
```

```python
# After — works reliably in 2.7.1
if emoji.id is not None:       # custom emoji (has a snowflake ID)
    lookup_str = emoji.id
else:                          # unicode emoji (id is None)
    lookup_str = emoji.name
```

This also applies to `reddit_publish.py` line 76:

```python
# Before
if payload.emoji.is_unicode_emoji() and payload.emoji.name == "📤":

# After
if payload.emoji.id is None and payload.emoji.name == "📤":
```

---

## Part 11: `.discriminator` usage

**Files: `pick_a_number.py` (line 108), `automod.py` (line 22)**

Discord migrated users away from the `username#1234` system in 2023. `member.discriminator` returns `"0"` for users on the new system. Replace any display usage with `member.name`.

**`pick_a_number.py`:**

```python
# Before
name = member.name + "#" + member.discriminator

# After
name = member.name
```

**`automod.py`:**

```python
# Before
await message.channel.send("**{}#{} ({}) was banned automatically.**\n{}.".format(
    message.author.name, message.author.discriminator, message.author.mention, reason
))

# After
await message.channel.send("**{} ({}) was banned automatically.**\n{}.".format(
    message.author.name, message.author.mention, reason
))
```

---

## Part 12: `poll_manager.py` — Monkey patch no longer needed

**File: `Anankos/poll_manager.py`**

The monkey patch added `message.poll` by extending `ComponentMessage`. With discord.py 2.4+, `message.poll` is a native attribute — no changes needed in `poll_manager.py` itself. The fix is solely the deletion of `monkey_patch.py` and its call in `bot.py` (covered in Parts 2 and 4).

Verify the poll data structure. The current code accesses `message.poll["question"]["text"]`. In discord.py 2.x, `message.poll` is a `discord.Poll` object, not a raw dict. Access the question text as `message.poll.question` instead:

```python
# Before (relying on monkey patch dict structure)
await create_thread(message, "Poll: " + message.poll["question"]["text"][:90], 4320)

# After (native Poll object)
await create_thread(message, "Poll: " + message.poll.question[:90], 4320)
```

---

## Summary: What's Compatible Without Changes

The following APIs are fully compatible with discord.py 2.7.1 and require no changes:

| API | Files |
|---|---|
| `discord.Embed`, `embed.set_image()`, `embed.add_field()`, etc. | `image_embed.py`, `trivia.py`, `reddit_publish.py` |
| `discord.File(buffer, filename=...)` | `pick_a_number.py`, `trivia.py`, `image_embed.py` |
| `message.channel.send()`, `message.reply()`, `message.edit()`, `message.delete()` | All modules |
| `message.add_reaction()`, `message.remove_reaction()` | `art_mention.py`, `reddit_publish.py` |
| `message.pin()`, `message.unpin()` | Any future usage |
| `message.author`, `message.guild`, `message.content`, `message.mentions`, `message.attachments` | All modules |
| `member.add_roles()`, `member.remove_roles()` | `role_reaction.py`, `permanent_roles.py`, `art_mention.py` |
| `guild.get_member()`, `guild.get_role()`, `guild.create_role()`, `guild.roles` | `art_mention.py` |
| `channel.webhooks()`, `webhook.send()` | `dragalia_notification.py`, `priconne_notification.py` |
| `channel.history()` | `art_mention.py` |
| `discord.AllowedMentions` | `art_mention.py` |
| `discord.Intents.all()` | `bot.py` — already correct |
| `client.is_closed()` | Background task loops (still exists in 2.7.1) |
| `client.wait_until_ready()` | Background task `before_loop` handlers |
| All event handler signatures (`on_ready`, `on_message`, `on_raw_reaction_add`, `on_member_join`, etc.) | `bot.py` |

---

## Migration Checklist

Work through these in order — each phase should leave the bot in a runnable state.

### Phase 1 — Dependencies and deleted files
- [ ] Update `discord.py==1.7.2` → `discord.py==2.7.1` in `requirements.txt`
- [ ] Remove `discord-py-slash-command` line from `requirements.txt`
- [ ] Replace `pycrypto==2.6.1` → `pycryptodome==3.20.0` in `requirements.txt`
- [ ] Delete `Anankos/monkey_patch.py`
- [ ] Run `pip install -r requirements.txt` and verify no install errors

### Phase 2 — Bot entry point and core structure
- [ ] Update `run.py` to use `asyncio.run(main())` with async context manager
- [ ] In `bot.py`: remove `discord_slash` and `monkey_patch` imports and the `monkey_patch()` call
- [ ] In `bot.py`: change `discord.Client` → `commands.Bot`, add `command_prefix` param
- [ ] In `bot.py`: remove `self.slash_command = SlashCommand(self)`
- [ ] In `bot.py`: replace `on_connect()` with `setup_hook()`
- [ ] In `bot.py`: replace `self.loop.create_task()` in `on_message` with `asyncio.create_task()`
- [ ] In `bot.py`: remove `on_component()` handler
- [ ] In `bot.py`: update `on_ready` print string (remove `#1337`)

### Phase 3 — Background tasks
- [ ] `trivia.py`: remove `self.client.loop.create_task()` from `__init__`, add `start_tasks()` + `@tasks.loop`
- [ ] `art_mention.py`: remove both `self.client.loop.create_task()` from `__init__`, add `start_tasks()` + `@tasks.loop` for both tasks
- [ ] `dragalia_notification.py`: same pattern, preserve minute-alignment logic in `before_loop`
- [ ] `priconne_notification.py`: same pattern
- [ ] `reddit_publish.py`: same pattern
- [ ] In `bot.py` `setup_hook()`: call `module.start_tasks()` for each module

### Phase 4 — Replace private HTTP API calls
- [ ] `utils.py`: replace `_state.http.request` with `await message.create_thread(name=..., auto_archive_duration=...)`
- [ ] `activities.py`: replace `_state.http.request` + `Invite.from_incomplete` with `await channel.create_invite(target_type=discord.InviteTarget.embedded_application, ...)`
- [ ] `reddit_publish.py`: replace `_state.http.request` crosspost with `await message.publish()`

### Phase 5 — Components rewrite (`art_mention.py`)
- [ ] Remove `from discord_slash.utils.manage_components import ...`
- [ ] Create `ArtMentionButton(discord.ui.DynamicItem)` class with `callback()`
- [ ] Replace `on_component()` with the `DynamicItem` callback
- [ ] Rewrite `on_message` button creation: `discord.ui.View` + `view.add_item(ArtMentionButton(...))`
- [ ] Rewrite `update_pingboard` button creation with same pattern
- [ ] In `bot.py` `setup_hook()`: add `self.add_dynamic_items(ArtMentionButton)`

### Phase 6 — Minor fixes
- [ ] `role_reaction.py`: replace `emoji.is_custom_emoji()` / `emoji.is_unicode_emoji()` with `emoji.id is not None` / `emoji.id is None`
- [ ] `reddit_publish.py`: same emoji check fix
- [ ] `pick_a_number.py`: remove `"#" + member.discriminator` from name formatting
- [ ] `automod.py`: remove `member.discriminator` from ban message
- [ ] `poll_manager.py`: update `message.poll["question"]["text"]` → `message.poll.question`

### Phase 7 — Testing
- [ ] Bot starts without errors
- [ ] `!subscribe`, `!unsubscribe`, `!listsubs` commands work
- [ ] Posting `!!charactername` in an image channel creates the notification message with buttons
- [ ] Clicking a subscribe button sends an ephemeral response
- [ ] Restart the bot — buttons on existing messages still work (persistent view test)
- [ ] Pingboard updates correctly with correctly-styled buttons
- [ ] `!activity` command works
- [ ] Reaction role assignment and removal work
- [ ] Reddit publish background task runs without errors
- [ ] Poll thread creation works
- [ ] Trivia background task starts and posts correctly
- [ ] Dragalia / Priconne notification background tasks start without errors

---

## Effort Estimate by File

| File | Change type | Effort |
|---|---|---|
| `requirements.txt` | Dependency swap | Trivial |
| `run.py` | Entry point pattern | Trivial |
| `monkey_patch.py` | Delete | Trivial |
| `bot.py` | Structural refactor — remove slash lib, add `setup_hook`, fix tasks | Medium |
| `art_mention.py` | Full component system rewrite + task migration | **High** |
| `utils.py` | One function, one API call | Trivial |
| `activities.py` | One function, one API call | Trivial |
| `reddit_publish.py` | Task migration + one crosspost call | Low |
| `trivia.py` | Task migration only | Low |
| `dragalia_notification.py` | Task migration only | Low |
| `priconne_notification.py` | Task migration only | Low |
| `pick_a_number.py` | String format fix | Trivial |
| `automod.py` | String format fix | Trivial |
| `role_reaction.py` | Two-line emoji check fix | Trivial |
| `poll_manager.py` | One attribute access fix | Trivial |

**Total: ~20–30 hours**, with the majority in `art_mention.py` (component rewrite) and `bot.py` (structural changes). The private HTTP API replacements in `utils.py`, `activities.py`, and `reddit_publish.py` are each single-function changes.
